Clean built extensions after install

eloyesp opened the pull request:
This should fix #3958, it is based on a comment of @simi there.

It just adds a `clean` step after `install` to remove object files.

## Make sure he following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

simi commented:
1. Some tests do rely on this.
2. I think it would be great to make this configurable somehow and start with this disabled by default.

deivid-rodriguez commented:
It's probably worth investigating the failing tests first before making a call on how to handle this. The only failing tests are some tests I added very recently to test rubygems custom require behaviour regarding `.rb` vs `.so` selection, and the issue might be perfectly with the tests themselves :shrug:

simi commented:
Any news in here?

deivid-rodriguez commented:
I just rebased this PR, and fix the test failure, which was indeed expected and doesn't signal any problem with doing this. I think doing this is safe.

deivid-rodriguez commented:
Oh wow, let me take back my previous message, this seems to break a lot of Bundler specs, actually. I will have a look.

eloyesp commented:
As far as I could research, there is one remaining issue in [`Gem::Ext::ExtConfBuilder#build`](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/ext/ext_conf_builder.rb#L13) that install the gem extensions within the gem itself by seting the `sitedir` argument for make, and making the `clean` task remove the actual installed files. The failure raises when the `full_tmp_dir` is being accessed after being cleaned up.

But I'm not sure why is ExtConfBuilder being used and why the sitedir is set to a tmpdir on the gem folder in that case.

eloyesp commented:
Ok, I now:

- fixed the failing tests properly (the ones checking the intermediate files instead of the installed ones).
- Added clean step on builder (that is called on install) that is quite simple to understand and have proper tests.
- Added an option `--dirty` that skip that step, just in case there is some gem broken by this change and added tests for it.
- Rebased everything on top of the `master` branch.

As far as I could run the tests everything just pass, but it seems that CI won't run without approval.

deivid-rodriguez commented:
Thanks for working on this @eloyesp, Since this would be your first contribution to this repo, CI needs manual approval. Just did that!

I'm not sure about `--dirty` option? I can't think of a case why someone would need artifacts to be left around, so I rather not commit to a new public CLI option that may not be used by anyone.

eloyesp commented:
The  `dirty` option is added because of a possible gem that may depend on the build artifacts, it could be replaced by an environment variable, the idea is simply to give an easy workaround for a broken gem.

Also, I were reviewing the failures and it seems that the cargo builder depends on the build artifacts (or at least the tests do), I cannot reproduce it here because I could not make cargo run here.

There is a single failure for bundler, but could not reproduce the issue (the test pass locally).

And there were two warnings because of an unused variable, I've fixed those.

deivid-rodriguez commented:
> As far as I could research, there is one remaining issue in [Gem::Ext::ExtConfBuilder#build](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/ext/ext_conf_builder.rb#L13) that install the gem extensions within the gem itself by seting the sitedir argument for make, and making the clean task remove the actual installed files. The failure raises when the full_tmp_dir is being accessed after being cleaned up.

> But I'm not sure why is ExtConfBuilder being used and why the sitedir is set to a tmpdir on the gem folder in that case.

I don't understand this case well, but I'd like to because using `make clean` feels... cleaner 😄. So, with the previous `make clean` approach, there was a single test failure, correct? Can you link push a separate branch with that approach and the failures it creates, I'd like to have a look.

eloyesp commented:
The problem with that approach is that the make method is a class method with many parameters, it is already doing too much and having lots of smells (in the [reek](https://github.com/troessner/reek) sense), adding the clean step there makes it much harder to debug and understand what happens.

While adding the single "clean" step in the array is clean enough adding that as an optional step adds lots of noise and making that cleaner would require a real refactor (or adding the seventh argument to that method).

simi commented:
I don't see problem with 7th optional argument. We need to maximize backward compatibility, this is the price we need to pay sometimes. Refactoring would be welcomed as well (not changing behaviour for now, just restructuring the code). All `ext` part of rubygems is a little confusing IMHO.

:thinking: I don't think `FileUtils.rm Dir["#{@spec.gem_dir}/**/*.{so,o,bundle}"]` is good approach, since extension can do actually a lot of different stuff (create various files) on different platforms (when extension can differ). What about linux static builds? `*.a` What about rules files `*.d`? What about Mac OS dynlibs aka `*.dylib`? We should go with `make clean` as standard task to cleanup environment. That is supported by `mkmf`, it seems to be supported by next release of rust `mkmf`, ...

eloyesp commented:
I understand that doing an actual removal with `FileUtils.rm` is not beautiful, but it have a clear advantage it is easier to understand and extend, that is this should get rid of all that you listed:

    FileUtils.rm Dir["#{@spec.gem_dir}/**/*.{so,o,bundle,a,d,dylib}"]

Refactoring everything on `ext` is actually an important task, but it is impossible for me (because I lack the time and the confidence with the code). The main issue with such a refactor is that there is a lot of coupling between the tests and the code, making any refactor require rewrite most of the test suite.

I were exploring the option of simply adding a "clean" step, and the least ugly way to get that working was using a global variable or something like that, that would make the needed refactor even harder to implement (adding more technical debt).

Once the refactor is actually done, I think that the best approach would be to use "make clean" as you all recomend, it is obviously cleaner that way.

simi commented:
I understand, but I still think we just can't remove random files we don't own in RubyGems. Those files are owned (created) by make tasks and they should be cleaned using the same owner.

For now, to keep it simple, what about to extend this line with post optional task?

https://github.com/rubygems/rubygems/blob/a09fdd10145d065d7f45917cd72a5a2f16720511/lib/rubygems/ext/builder.rb#L43

```ruby
["clean", "", "install", ENV['RUBYGEMS_POST_MAKE_TASK']].compact.each do |target|
```

Would that be enough for now for opt-in behaviour of running `make clean` doing something like `RUBYGEMS_POST_MAKE_TASK=clean gem install xyzl`? Would you mind to give it a try?

eloyesp commented:
Ok, I've created another pull-request with that implementation, it uses an ENV variable (that is another way to use global state) making testing and refactoring even harder.

deivid-rodriguez commented:
From what I understand, there's two discussions going on:

* Whether to `make clean` or remove specific files. I agree with @simi here.
* Whether to make this opt in, or not. I'm not necessarily against making this opt-in, but I don't know of any gem relying on intermediate build artifacts present in the final installation directory. Maybe one such gem exists, but I tend towards potentially breaking some users of this potential gem in favor of benefiting the rest of the community by default. Apparently the cargo builder relies on intermediate build artifacts being present in the installation folder? Can we look a bit more into that? Maybe it's just a problem with the tests?

eloyesp commented:
Yes and no, I do think as everyone here that using "make clean" is better than removing files using a glob, but I prefer the last given the current implementation of the extension building process complexity. That is, adding more complexity there adds more technical debt and makes it even harder to refactor the code.

So, I'm not against using "make clean", I'm against the environment variable. That is the trade-off I'm talking about.

Regarding the opt-in or opt-out, I also think that opt-out is the way to go, but it is quite important to make that opt-out very clear, that is why I've added the `dirty` flag there, it sounds bad.

simi commented:
> Ok, I've created another pull-request with that implementation, it uses an ENV variable (that is another way to use global state) making testing and refactoring even harder.

That was just a quick idea to roll it out quickly for testing. Later we can integrate this in better way and turn it opt-out (enabled) by default. If you don't like this idea, we can continue exploring.

simi commented:
@deivid-rodriguez cargo/ruby connection did this recently -https://github.com/oxidize-rb/rb-sys/pull/107.

eloyesp commented:
Ok, I'm almost sure I found the bug with cargo builder... It seems that when installing it adds a "release" subfolder that makes require fail.

```
# doing this manually in the middle of the test make it pass
cp tmp/test_rubygems_20221201-5322-h1rss4/gemhome/extensions/x86_64-linux/3.1.0-static/rust_ruby_example-0.1.0/release/rust_ruby_example.so tmp/test_rubygems_20221201-5322-h1rss4/gemhome/extensions/x86_64-linux/3.1.0-static/rust_ruby_example-0.1.0/
```

I'm working on a proper fix.

eloyesp commented:
That last commit should fix most of the failing tests (I could not reproduce the bundler related failure locally so I hope this fix that issue).

Now, we can:

- leave that clean task as is, and improve it in a later iteration with more information.
- add a little bit of love to that glob with the information we already have.
- try different strategies like running "make clean", something like a partial pristine (remove anything not listed in the gem itself). (with the current implementation this would require adding ~50 LOC)

My recommendation would be the first or the second one, leaving the improvements to a later iteration. I think that after a proper refactor the `make clean` option should be a one-liner, but adding it now will just make that refactor even harder (as it will need to duplicate most of the `build_extensions` related code.)

eloyesp commented:
This extra line makes the test fail in a much clear way:

``` ruby
run 'puts $LOAD_PATH.grep(/extensions/).each { |path| raise "Invalid LOAD_PATH: #{ path }
  " unless File.directory?(path) }'
```

eloyesp commented:
Found the issue, but I will not be able to fix it in the near future. The problem is that bundler when looking for performance is breaking the DRY rule duplicating the Specification using different implementations on different stages.

While the gem is being installed the extensions_path is defined in [rubygems_ext.rb](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/rubygems_ext.rb#L67) monkeypatching Gem::Specification.

But afterwards on [runtime](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/runtime.rb#L24) those specs are [a special collection](https://github.com/rubygems/rubygems/blob/771c94e14d6c86c8df3d1df0d2679ed59316252a/bundler/lib/bundler/spec_set.rb) of [stub specifications](https://github.com/rubygems/rubygems/blob/771c94e14d6c86c8df3d1df0d2679ed59316252a/bundler/lib/bundler/stub_specification.rb) that should not be confused with [stub specifications](https://github.com/rubygems/rubygems/blob/771c94e14d6c86c8df3d1df0d2679ed59316252a/lib/rubygems/stub_specification.rb).

The problem is that there is a mismatch regarding what the `extension_dir` actually is.

I've been digging several hours in this, but I cannot fix that as it requires much more knowledge about how to debug this.

eloyesp commented:
I mostly sure I fixed the bundler bug, but it would need a proper refactor to reuse instead of duplicate. I've just changed the stub implementation to match.

eloyesp commented:
It seems that _some_ extensions are being installed in different paths, so now that I modify the "extension_path" to make it work for one gem, it is broken for other gem, may be there is a better way to modify the "extension_path" to make everyone happy.

eloyesp commented:
I'm still working on this bug, it seems that bundler installs git sourced gems extensions on two different paths depending something (not sure what).

The two cases are on the specs, but they are not specified, so the specs pass because ruby finds the build artifacts (`$LOAD_PATH` is set to a non-existing directory but that is ok for the specs).

The test on `spec/install/gems/native_extensions_spec.rb:93` install extensions on something like:

    #{ stub.base_dir }/#{ source.extension_dir_name }/extensions/etc

While the test `./spec/install/gemfile/git_spec.rb:1203` installs the extensions on:

    #{ stub.base_dir }/extensions/etc

I could not find where is the difference, what makes this case use different folder structure. But without fixing this bug, adding the `clean artifacts` feature on rubygems might break some bundler setups with git gems.

I would greatly appreciate some help here. Regards.

deivid-rodriguez commented:
@eloyesp I'll try to have a look and see if I can figure this out. Thanks so much for your persistence on this issue, it's of great help!

deivid-rodriguez commented:
Closing in favor of #6133, let's continue the discussion there.
